### DISCUSSION TITLE
Any known viewer which is able to help determin what BREP data is available?

### DISCUSSION THREAD

**Discussion by Mannshoch** [IfcOpenShell/IfcOpenShell#1724] [Q&A]
I'm Working in the Industry with a Proprietary CAD system. More and More times we get IFC files that contains the whole Building instead of the geometry and internals of a certain room.
The available IFC viewers only shows IFC and Tessellation data but not what BREP is available.

To be able to work with the Data we need BREP data. We need a fast way to determine if certain Data in the IFC not only has Tesselation data but also BREP data. 

Does someone know a Viewer that let me show what geometry is available as BREP?

**Comment by Moult** [comment]
The BlenderBIM Add-on will show you the geometric representation types available for each object. It will tell you the context, as well as the geometry type.

Shown in this screenshot, there is a body made from a swept solid, and a footprint using a geometric curve set.

It will distinguish between Breps and Tessellations.

![2021-09-06-162924_654x124_scrot](https://user-images.githubusercontent.com/88302/132170659-db5c6868-fa98-42e6-9b11-0799d2f5476c.png)

**Reply by Mannshoch** [reply]
I'll test that. But Blender is not really a candidate for a tiny and easy to use IFC viewer.

**Reply by aothms** [reply]
@Mannshoch it depends. It boots quickly and the UI is fully customizable. There are quite a few tools that use Blender but hide the complexity in the UI.

Alternatively you create a python script with a UI, a bit like https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcopenshell-python/ifcopenshell/geom/app.py#L575

**Reply by Mannshoch** [reply]
Enhancement request: 
Fork Blender with a custom UI and release it as an IFC-Viewer. 😁

I know Blender but never used it before. I'm also not a developer.
Installing Blender worked, also installing the add-in. But I was only able to import an IFC into a new Blender file and not open it directly.

**Reply by Mannshoch** [reply]
@Moult I'm Playing around with Blender. How could I filter the screen for showing only BREP data?
![grafik](https://user-images.githubusercontent.com/7795891/132294821-fed6b10b-66fc-4706-abef-a28f2d741da0.png)

**Reply by aothms** [reply]
>  But I was only able to import an IFC into a new Blender file and not open it directly.

This is something I have also thought about in the past. Making Blender a general purpose file viewer.

If you have a .bat script somewhere with something like

`blender -P open.py -- %*`

and a file open.py with something like:

```
import bpy
import sys
sep = sys.argv.index('--')
for fn in sys.argv[sep + 1:-1]:
    if fn.endswith(".obj"):
        bpy.ops.import_scene.obj(filepath=fn, axis_forward='Y', axis_up='Z', use_split_groups=True)
    elif fn.endswith(".dae"):
        bpy.ops.wm.collada_import(filepath=fn)
    elif fn.endswith(".ifc"):
        #invoke blender bim functionality
        pass
```

I think you can be actually quite close. I had something like that in the past on an old computer, but couldn't find it anymore.

Then you can have the .bat script as the default action to open a file or under your open with ... menu.

**Reply by Moult** [reply]
Unfortunately there is no graphical button to filter object on the screen by representation type. You can use the built in Python interpreter to quickly do this and I can show you how, but is that what you want?

From a technical standpoint, you are actually opening an IFC directly. Even though it's called "Import" and in the "Import" menu, it's actually only creating a cache of geometry and materials in Blender, but all the raw data is read directly from IFC. In that sense, it is exactly like "Open" as you would expect a native file to work. The Blender file, if you choose to save it, is semantically meaningless, just a convenient geometry cache. Similarly, when you see "Export", it's not actually an export, it's more like a "Save". These names "Import" and "Export" are a bit historical back from the first prototypes of the BlenderBIM Add-on existed and it did function as an Import and Export, but they are no longer accurate.

From a UX standpoint, yes, it would be nice to make it more "viewer" like but this hasn't been high up on the priority list unfortunately.

**Reply by Mannshoch** [reply]
> Unfortunately there is no graphical button to filter object on the screen by representation type. You can use the built in Python interpreter to quickly do this and I can show you how, but is that what you want?

Not really. May I'm able to follow instructions and get the result I need but I'm not looking for a solution for me alone. In fact, I'm not the designated user of the IFC Viewer I search.

**Comment by Mannshoch** [comment]
Created a Feature Request:
https://github.com/IfcOpenShell/IfcOpenShell/issues/1726

**Reply by Mannshoch** [reply]
@aothms I would like to see Blender as a FOSS 3D-Viewer Software for Windows and Linux. The Interesting point is that the whole Blender Software (640MB) is smaller than my current viewer for 3D Files.

**Reply by aothms** [reply]
Are you interested in reaching out to that community? I.e blenderartists.org and also Ton himself is really approachable.

**Reply by Mannshoch** [reply]
Until this installation I never worked with Blender. I found this Project be searching google for Information about IFC. I'm not a developer. On Github I only write Issues, Enhancement requests and sometimes Mockups for FOSS that I like. I'm not sure If I'm the right Person for that.

**Reply by aothms** [reply]
It's not only about coding, it's about enthusing others, getting an understanding of the complexity, similar existing efforts, ...

**Reply by Mannshoch** [reply]
Seems interesting. 
I simply startet a request her: 
https://blenderartists.org/t/any-possibilities-to-have-blender-as-a-file-viewer/1326721
not sure if correct that way but lets see what happens.

**Reply by Moult** [reply]
There is a way to make it visually as simple as a viewer. This uses whats called an application template, and would likely also use nothing but the 3D view. Andrew Peel (Homebuilder dev) has done quite a bit of work on this. The bigger worry I have is that 1) IfcOpenShell is still slower than other viewers when it comes to brep data and 2) Blender is slower than other viewers when it comes to large numbers of objects. How large are your models (perhaps better measured in number of objects, not filesize)? Are they less than 30,000 objects (small), or greater than 70,000 (large)?

I will see what I can do about providing this UX. Got a few TODO items lined up, but I'll get around to it.

**Reply by aothms** [reply]
Yes that's a good point, there is a guaranteed overhead in keeping the meshes editable as is the case in Blender.

**Reply by Mannshoch** [reply]
The speed of Blender in viewing IFC is not the fastest but acceptable. I do not know how many parts these IFC have. I don't know how to count them.

Thanks for your work!
